metadata/src/main/java/org/apache/kafka/controller/QuorumController.java

```diff
@@ -1200,6 +1235,16 @@ public static List<ApiMessageAndVersion> generateActivationRecords(
                     throw new RuntimeException("Should not have ZK migrations enabled on a cluster running metadata.version " + featureControl.metadataVersion());
                 }
             }
+
+            if (inTransaction) {
+                if (!featureControl.metadataVersion().isMetadataTransactionSupported()) {
+                    throw new RuntimeException("Detected in-progress transaction, but the metadata.version " + featureControl.metadataVersion() +
+                        " does not support transactions. Cannot continue.");
+                } else {
+                    log.warn("Detected in-progress transaction during controller activation. Aborting this transaction.");
```

Comment:
I think we should put the start transaction offset (if you want, pass an OptionalLong rather than boolean?) That helps match up the beginning of the transaction with where it ends.

Also should fill in that "reason" field, presumably with something like "controller failover"